src/gangliad/statsd.cpp

```diff
@@ -863,7 +880,45 @@ StatsD::ParseMetrics( std::string const &stats_metrics_string, char const *param
 		StringList target_types(target_type.c_str());
 		m_target_types.create_union(target_types,true);
 
+		// Add this metric ad to our list of metrics
 		stats_metrics.push_back(ad);
+
+		// For even more efficient queries to the collector, keep track of 
+		// which ad attributes we need (attribute projection).
+		// Note that we need to give up on using a projection if:
+		//    1. any metric has a RegEx attribute
+		//    2. any metric does not have a value  AND has a name that is not a string literal
+		// Happily, both of the above conditions are pretty advanced and thus rarely used.
+		if (!m_want_projection) continue;
+		if (ad->Lookup(ATTR_REGEX)) {
+			// Crap, we have to bail on using a projection
+			dprintf(D_FULLDEBUG,"No collector projection: metric found with %s attribute\n",ATTR_REGEX);
+			m_want_projection = false;
+			continue;
+		}
+		if (!ad->Lookup(ATTR_VALUE)) {
+			std::string attr_name;
+			if (ExprTreeIsLiteralString(ad->Lookup(ATTR_NAME),attr_name)) {
+				// add this to projection list
+				m_projection_references.insert(attr_name);
+			} else {
+				// Crap, we have to bail on using a projection
+				dprintf(D_FULLDEBUG,
+					"No collector projection: metric found without a Value attribute and a non-literal Name\n");
+				m_want_projection = false;
+				continue;
+			}
+		}
+		classad::ClassAd::const_iterator attr_it;
```

Comment:
Is this `const_iterator` variable needed?

Reply:
Nope.  Good catch, wonder why the compiler did not give a warning. It is now removed on this branch.